### DISCUSSION TITLE
How can I log request and response with "smallrye-graphql-client-implementation-vertx" ?

### DISCUSSION THREAD

**Discussion by jmini** [smallrye/smallrye-graphql#2207] [General]
I am currently trying the typesafe client with `smallrye-graphql-client-implementation-vertx` instantiated from a simple main class:

```
SuperheroesClientApi api = TypesafeGraphQLClientBuilder.newBuilder()
                                                        .build(SuperheroesClientApi.class);
```

Source: [Client.java](https://github.com/jmini/quarkus-experiments/blob/8b95fbb3741094629bb9eba4fbb4d3f0e6dd1fbf/quarkus43756_smallrye-graphql-client/src/main/java/run/Client.java#L17)

---

In Quarkus I know that there is an option to log client and response:

```
quarkus.log.category."io.smallrye.graphql.client".level=TRACE
quarkus.log.category."io.smallrye.graphql.client".min-level=TRACE
```

---

Is there something similar in `io.smallrye:smallrye-graphql-client-implementation-vertx`, and how can I enable this?

_I have tried to search for modifying log level `org.jboss.logging:jboss-logging` but the search results are pointing to changing log levels inside the JBoss app server. I was not able to understand how it can be configured programmatically_

**Comment by jmartisk** [comment]
I'd actually be interested in learning this too. Maybe @jamezp can help us -  what is the easiest way to get a standalone (Java SE) application that uses the JBoss Logging API to actually log something? It's hard to find any documentation that is not specific to WildFly or Quarkus.

**Reply by jmini** [reply]
^ Very good summary of my question 👍

**Reply by jamezp** [reply]
JBoss Logging is just a logging facade. If you want a log manager you'd have to use something like JUL or JBoss Log Manager. I can give some instructions for that if it helps.

**Reply by t1** [reply]
A link to some documentation would be wonderful 😁 AFAICS, the JBoss Log Manager is a nice, mature, and powerful piece of software. It only lacks publicly available documentation.

**Reply by jmartisk** [reply]
I've submitted https://github.com/smallrye/smallrye-graphql/pull/2209 to provide a simple example in our docs

**Comment by jmini** [comment]
Ok I found everything I needed:

**JBoss Logging** is the facade (like `slf4j`):
https://github.com/jboss-logging/jboss-logging
--> `smallrye-graphql-client-implementation-vertx` has already a dependency on it.

By default it seems to do something, but I am not sure if this is configurable and how.

**JBoss Log Manager** can control what is logged and where (like `logback`)
https://github.com/jboss-logging/jboss-logmanager

* It has to be added on the runtime classpath
* It has to be enabled with `-Djava.util.logging.manager=org.jboss.logmanager.LogManager` when the Java SE process is started
* The configuration file `logging.properties` can be in the jar, so a simple way is to add it to `src/main/resources`

About the **configuration file**:

I found this example: [`logging.properties` from jamezp/jboss-logging-example](https://github.com/jamezp/jboss-logging-example/blob/master/src/main/resources/logging.properties)

I simplified it (I removed what I did not understood) and I managed to enable:

* DEBUG level for my main class: `run.Client`
* TRACE level for for `io.smallrye.graphql.client` (similar to what is recommended by quarkus)

And 🎉 :
https://github.com/jmini/quarkus-experiments/commit/7e357617921bfc9b02e2c95352ca35c0fa1d84a2

`./gradlew run`

```
> Task :run
13:17:00,422 run.Client.main:16 (main)  INFO Client:16 - --- Start ---
13:17:00,497 run.Client.main:17 (main) DEBUG Client:17 - Timestamp: 1728645420494
13:17:00,834 io.smallrye.graphql.client.vertx.VertxManager.getOrCreateCustom:84 (main) DEBUG VertxManager:84 - Using custom Vert.x instance io.vertx.core.impl.VertxImpl@65fb9ffc
13:17:00,973 io.smallrye.graphql.client.vertx.typesafe.VertxTypesafeGraphQLClientProxy.request:314 (main) TRACE VertxTypesafeGraphQLClientProxy:314 - full graphql request: {"query":"query allCities { allCities {name symbol} }","variables":{},"operationName":"allCities"}
13:17:01,258 io.smallrye.graphql.client.vertx.typesafe.VertxTypesafeGraphQLClientProxy.executeSingleResultOperationOverHttpSync:174 (main) TRACE VertxTypesafeGraphQLClientProxy:174 - response graphql: {"data":{"allCities":[{"name":"Gotham City","symbol":"GC"},{"name":"New York City","symbol":"NYC"}]}}
response: [City [name=Gotham City, symbol=GC], City [name=New York City, symbol=NYC]]
13:17:01,302 io.smallrye.graphql.client.vertx.typesafe.VertxTypesafeGraphQLClientProxy.request:314 (main) TRACE VertxTypesafeGraphQLClientProxy:314 - full graphql request: {"query":"query allSuperheroes { allSuperheroes {abilities{__typename ... on Attribute {motivation name rating} ... on Superpower {name}} city {name symbol} gender name} }","variables":{},"operationName":"allSuperheroes"}
13:17:01,312 io.smallrye.graphql.client.vertx.typesafe.VertxTypesafeGraphQLClientProxy.executeSingleResultOperationOverHttpSync:174 (main) TRACE VertxTypesafeGraphQLClientProxy:174 - response graphql: {"data":{"allSuperheroes":[{"abilities":[{"__typename":"Superpower","name":"Genius-Level Intellect"},{"__typename":"Superpower","name":"Exp....
```

**Reply by jmartisk** [reply]
Nice @jmini, thanks for digging into it! I'll add something like this to the SmallRye GraphQL docs!

**Comment by jmini** [comment]
There is even a Jbang way of doing what I have described in https://github.com/smallrye/smallrye-graphql/discussions/2207#discussioncomment-10914249:

See commit: https://github.com/jmini/gitlab-experiments/commit/8602b2c4626945c8d11a1256eab9a3942758c03e
